File: dotnet/core/controller/PersonController.cs
using Newtonsoft.Json;
using core.dao;
using core.model;
using core.mapper;

namespace core.controller;

public class PersonController
{
    private static readonly string GET = "GET";

    private static readonly string POST = "POST";

    private readonly PersonDao _personDao;

    public PersonController(PersonDao personDao)
    {
        _personDao = personDao;
    }

    public async Task<GenericResponse> HandleRequest(GenericRequest request)
    {
        var response = new GenericResponse();
        
        if (GET.Equals(request.HttpMethod) && "/person".Equals(request.Path))  
		{
			var idQueryParameters = request.QueryParameters["id"];
            if (idQueryParameters.Count == 0) 
            {
                throw new NullReferenceException("id is a required parameter");
            }

            var id = idQueryParameters[0];
            var uuid = Guid.Parse(id);

            var person = await _personDao.GetById(uuid);
            Console.WriteLine("retrieved person is: {0}", person);
            if (person != null) 
            {
                response.Status = 200;

                var personJson = JsonConvert.SerializeObject(person);
                Console.WriteLine(personJson);
                
                response.Body = personJson;
            }
            else 
            {
                response.Status = 404;
                return BeforeResponse(request, response);
            }
		}
        else if (POST.Equals(request.HttpMethod) && "/person".Equals(request.Path))
        {
            if (request.Body == null) {
                response.Status = 400;
                return BeforeResponse(request, response);
            }
            
            Console.WriteLine("request body before deserialization is: {0}", request.Body);
            try
            {
                var person = JsonConvert.DeserializeObject<Person>(request.Body);
                Console.WriteLine("person uuid after deserialization is: {0}", person.id);
                
                var result = await _personDao.Save(person);
                Console.WriteLine("result is: {0}", result);
                Console.WriteLine("person uuid after saving is: {0}", person.id);

                response.Status = 201;
                response.Body = JsonConvert.SerializeObject(
                    new Dictionary<string, string>{
                        { "id", person.id.ToString() }
                    }
                );
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                throw e;
            }
        }
        else 
        {
            Console.WriteLine("No endpoint matches {0} and {1}", request.HttpMethod, request.Path);
            response.Status = 404;
        }

        return BeforeResponse(request, response);
    }

    private static GenericResponse BeforeResponse(GenericRequest request, GenericResponse response)
    {
        response.AddHeader("Content-Type", "application/json");
        return response;
    }
}

File: typescript/core/src/controller/person-controller.ts
import {PersonDao} from "../dao/person-dao";
import {GenericRequest} from "../mapper/generic-request";
import {Optional} from "typescript-optional";
import {Person} from "../model/person";
import {GenericResponse} from "../mapper/generic-response";

export class PersonController {

    private static readonly GET: string = 'GET';

    private static readonly POST: string = 'POST';

    // todo:  do we need something like Gson in Java to handle Date serialization and deserialization?

    private readonly personDao: PersonDao;

    constructor(personDao: PersonDao) {
        this.personDao = personDao;
    }

    public async handleRequest(request: GenericRequest, response: GenericResponse) {
        if (PersonController.GET === request.httpMethod && '/person' === request.path) {
            console.log(`query parameters: ${JSON.stringify(request.queryParameters)}`);

            const id: string = Optional.ofNullable(request.queryParameters['id'])
                .orElseThrow(() => new Error('id parameter is required'));

            const personOptional: Optional<Person> = await this.personDao.getById(id);
            personOptional.ifPresentOrElse(
                person => {
                    console.log(person);

                    response.statusCode = 200;

                    const personJson: string = JSON.stringify(person);
                    console.log(personJson);

                    response.body = personJson;
                },
                () => {
                    response.statusCode = 404;
                }
            )
        }
        else if (PersonController.POST === request.httpMethod && '/person' === request.path) {
            const person: Person = request.body as Person;

            await this.personDao.save(person);

            response.statusCode = 201;
            response.body = JSON.stringify({
                'id': person.id
            })
        }
        else {
            console.log(`No endpoint matches ${request.httpMethod} ${request.path}`)
        }

        PersonController.beforeResponse(request, response);
    }

    /**
     * This method contains logic that should run AFTER the request has been handled, but BEFORE the response is sent back
     * to the client.
     *
     * @param request {@link GenericRequest}
     * @param response {@link GenericRequest}
     * @private
     */
    private static beforeResponse(request: GenericRequest, response: GenericResponse): void {
        response.addHeader('Content-Type', 'application/json');
    }

}